modules/tinymce/src/core/main/ts/api/commands/ClipboardCommands.ts

```diff
@@ -1,38 +1,102 @@
+import { PlatformDetection } from '@ephox/sand';
+
 import Editor from '../Editor';
 import Env from '../Env';
 
 export const registerCommands = (editor: Editor): void => {
-  editor.editorCommands.addCommands({
-    'Cut,Copy,Paste': (command) => {
-      const doc = editor.getDoc();
-      let failed;
+  const browser = PlatformDetection.detect().browser;
+
+  const copyToClipboard = async (text: string) => {
+    await navigator.clipboard.writeText(text);
+  };
+
+  const pasteFromClipboard = async (): Promise<void> => {
+    let text = '';
+    text = await navigator.clipboard.readText();
+    editor.selection.setContent(text);
+  };
+
+  const cutToClipboard = async (text: string) => {
+    await navigator.clipboard.writeText(text);
+    editor.selection.setContent('');
+  };
+
+  const handleSecureContentError = () => {
+
+    let msg = editor.translate(
+      'Error: This operation requires a secure context (HTTPS)' +
+        'Direct access to the clipboard is not allowed in insecure contexts.'
+    );
+
+    if (Env.os.isMacOS() || Env.os.isiOS()) {
+      msg = msg.replace(/Ctrl\+/g, '\u2318+');
+    }
+
+    editor.notificationManager.open({ text: msg, type: 'error' });
+
+    return false;
+  };
 
-      // Try executing the native command
+  const checkPermission = async () => {
+    const permission = browser.isChromium() ? await navigator.permissions.query({ name: 'clipboard-read' as PermissionName }) : { state: 'granted' };
+    if (permission.state === 'granted') {
+      return true;
```

Comment:
Now that you've rewritten this file, these checks are already in the only code that calls this function. Maybe the `denied` state should move there too?

- [ ] de-duplicate and simplify